core/vm/state_manager.go

```diff
@@ -71,24 +39,16 @@ func methodSignatureToMethodID(methodSignature string) [4]byte {
 	return methodID
 }
 
-func stateManagerRequiredGas(input []byte) (gas uint64) {
-	var methodID [4]byte
-	copy(methodID[:], input[:4])
-	gas = methodIds[methodID].smGasCost
-	return gas
-}
-
 func callStateManager(input []byte, evm *EVM, contract *Contract) (ret []byte, err error) {
 	var methodID [4]byte
+	if len(input) == 0 {
+		return nil, nil
+	}
 	copy(methodID[:], input[:4])
-	ret, err = methodIds[methodID].smFunction(evm, contract, input)
+	ret, err = methodIds[methodID](evm, contract, input)
```

Comment:
Since `methodID` is dynamic, this feels a bit safer:

```go
if method, ok := methodIds[methodID]; ok {
    ret, err = method(evm, contract, input)
    return ret, err
}

return nil, errors.New("Method not found")
```

I don't love the way that the error case of "method not found" isn't returned immediately. To fix this we would check `!ok` in the `if` clause but then `method` isn't in context outside of the `if` block.

Reply:
You could define `method` above the check so that it is in scope afterwards. What do you think?